//
// Copyright (C) 2024 EA group inc.
// Author: Jeff.li lijippy@163.com
// All rights reserved.
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
//


syntax="proto2";

package melon.policy;

option cc_generic_services = true;
option java_generic_services = true;
option java_package="com.melon.policy";
option java_outer_classname="MongoProto";

enum MongoOp {
    OPREPLY = 1;
    DBMSG = 1000;
    DB_UPDATE = 2001;
    DB_INSERT = 2002;
    DB_QUERY = 2004;
    DB_GETMORE = 2005;
    DB_DELETE = 2006;
    DB_KILLCURSORS = 2007;
    DB_COMMAND = 2008;
    DB_COMMANDREPLY = 2009;
}

message MongoHeader {
    required int32 message_length = 1;
    required int32 request_id = 2;
    required int32 response_to = 3;
    required MongoOp op_code = 4;
}

message MongoRequest {
    required MongoHeader header = 1;
    required string message = 2;
}

message MongoResponse {
    required MongoHeader header = 1;
    required int32 response_flags = 2;
    required int64 cursor_id = 3;
    required int32 starting_from = 4;
    required int32 number_returned = 5;
    required string message = 6;
}

service MongoService {
    rpc default_method(MongoRequest) returns (MongoResponse);
}
